compiler-rt/lib/scudo/standalone/checksum.cpp

```diff
@@ -75,6 +77,15 @@ bool hasHardwareCRC32() {
   return !!(getauxval(AT_HWCAP) & HWCAP_CRC32);
 #endif // SCUDO_FUCHSIA
 }
+#elif defined(__loongarch__)
+// Query HWCAP for platform capability, according to *Software Development and
+// Build Convention for LoongArch Architectures* v0.1, Section 9.1.
+//
+// Link:
+// https://github.com/loongson/la-softdev-convention/blob/v0.1/la-softdev-convention.adoc#kernel-development
+bool hasHardwareCRC32() {
+  return !!(getauxval(AT_HWCAP) & HWCAP_LOONGARCH_CRC32);
```

Comment:
I got this error:
```
compiler-rt/lib/scudo/standalone/checksum.cpp:87:35: error: use of undeclared identifier 'HWCAP_LOONGARCH_CRC32'
```

Maybe a `.h` file should be included?

Reply:
After `<asm/hwcap.h>` is included, I got this error:
```
compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:49: error: implicit conversion changes signedness: 'u32' (aka 'unsigned int') to 'int' [-Werror,-Wsign-conversion]
   23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
      |                           ~~~~~~~~~~~~~~~       ^~~
compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:43: error: implicit conversion changes signedness: 'uptr' (aka 'unsigned long') to 'long' [-Werror,-Wsign-conversion]
   23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
      |                           ~~~~~~~~~~~~~~~ ^~~~
2 errors generated.
```

Note that I'm using clang-17 to build.

Reply:
> After `<asm/hwcap.h>` is included, I got this error:
> 
> ```
> compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:49: error: implicit conversion changes signedness: 'u32' (aka 'unsigned int') to 'int' [-Werror,-Wsign-conversion]
>    23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
>       |                           ~~~~~~~~~~~~~~~       ^~~
> compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:43: error: implicit conversion changes signedness: 'uptr' (aka 'unsigned long') to 'long' [-Werror,-Wsign-conversion]
>    23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
>       |                           ~~~~~~~~~~~~~~~ ^~~~
> 2 errors generated.
> ```
> 
> Note that I'm using clang-17 to build.

Yes, this is nasty.  `__crc_w_d_w` etc. has signed argument and return value types.  I had to add a lot of casts in my xz CRC32 optimization to silence the warning.

Reply:
> I got this error:
> 
> ```
> compiler-rt/lib/scudo/standalone/checksum.cpp:87:35: error: use of undeclared identifier 'HWCAP_LOONGARCH_CRC32'
> ```
> 
> Maybe a `.h` file should be included?

Hmm I was wrong in assuming the HWCAP constants are implicitly pulled in by `<sys/auxv.h>`. This is the case only [from glibc 2.38](https://github.com/bminor/glibc/commit/7f079fdc16e88ebb8020e17b2fd900e8924da29a). I'll supply the definition with an `#ifdef` check.